jina/executors/indexers/vector.py

```diff
@@ -69,10 +70,68 @@ def __init__(self,
                                 f'compress_level overriden from `ref_indexer` to {ref_indexer.compress_level} \n'
                                 f'index_filename overriden from `ref_indexer` to {ref_indexer.index_filename}')
             self.ref_indexer_workspace_name = ref_indexer.workspace_name
+            self.delete_on_dump = getattr(ref_indexer, 'delete_on_dump', delete_on_dump)
+
+
+    def _delete_invalid_indices(self):
+        valid = self.valid_indices[self.valid_indices == True]  # noqa
+        if len(valid) != len(self.valid_indices):
+            self._clean_memmap()
+            self._post_clean_memmap(valid)
+
+    def _post_clean_memmap(self, valid):
+        # here we need to make sure the fields
+        # that depend on the valid_indices are cleaned up too
+        valid_key_bytes = np.frombuffer(self.key_bytes, dtype=(np.str_, self.key_length))[
+            self.valid_indices].tobytes()
+        self.key_bytes = valid_key_bytes
+        self._size = len(valid)
+        self.valid_indices = valid
+        del self.__dict__['CACHED__int2ext_id']
+        del self.__dict__['CACHED__ext2int_id']
+
+    def _clean_memmap(self):
+        # clean up the underlying matrix of entries marked for deletion
+        # first we need to make sure we flush the writing handler
+        if self.write_handler and not self.write_handler.closed:
+            with self.write_handler as f:
+                f.flush()
+                f.close()
```

Comment:
when using a file as context manager, close is not required to be called explicitly